src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -441,6 +453,201 @@ try
+}
+
+// Routine Description:
+// - Updates our internal map of font features with the given features
+// - NOTE TO CALLER: Make sure to call UpdateFont after calling this for the feature changes
+//   to take place
+// Arguments:
+// - features - the features to update our map with
+void DxFontRenderData::SetFeatures(const std::unordered_map<std::wstring_view, uint32_t>& features)
+{
+    // Populate the feature map with the standard list first
+    std::unordered_map<DWRITE_FONT_FEATURE_TAG, uint32_t> featureMap{
+        { DWRITE_MAKE_FONT_FEATURE_TAG('r', 'l', 'i', 'g'), 1 }, // Required Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('r', 'c', 'l', 't'), 1 }, // Required Contextual Alternates
+        { DWRITE_MAKE_FONT_FEATURE_TAG('l', 'o', 'c', 'l'), 1 }, // Localized Forms
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'c', 'm', 'p'), 1 }, // Glyph Composition / Decomposition
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'a', 'l', 't'), 1 }, // Contextual Alternates
+        { DWRITE_MAKE_FONT_FEATURE_TAG('l', 'i', 'g', 'a'), 1 }, // Standard Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'l', 'i', 'g'), 1 }, // Contextual Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('k', 'e', 'r', 'n'), 1 }, // Kerning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'a', 'r', 'k'), 1 }, // Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'k', 'm', 'k'), 1 }, // Mark to Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('d', 'i', 's', 't'), 1 }  // Distances
+    };
+
+    // Update our feature map with the provided features
+    if (!features.empty())
+    {
+        for (const auto [tag, param] : features)
+        {
+            if (tag.length() == TAG_LENGTH)
+            {
+                featureMap.insert_or_assign(DWRITE_MAKE_FONT_FEATURE_TAG(til::at(tag, 0), til::at(tag, 1), til::at(tag, 2), til::at(tag, 3)), param);
+            }
+        }
+        _didUserSetFeatures = true;
+    }
+
+    // Convert the data to DWRITE_FONT_FEATURE and store it in a vector for CustomTextLayout
+    _featureVector.clear();
+    for (const auto [tag, param] : featureMap)
+    {
+        _featureVector.push_back(DWRITE_FONT_FEATURE{ tag, param });
+    }
+}
+
+// Routine Description:
+// - Updates our internal map of font axes with the given axes
+// - NOTE TO CALLER: Make sure to call UpdateFont after calling this for the axes changes
+//   to take place
+// Arguments:
+// - axes - the axes to update our map with
+void DxFontRenderData::SetAxes(const std::unordered_map<std::wstring_view, int32_t>& axes)
+{
+    _axesVector.clear();
+
+    // Update our axis map with the provided axes
+    for (const auto [axis, value] : axes)
+    {
+        if (axis.length() == TAG_LENGTH)
+        {
+            const auto dwriteTag = DWRITE_MAKE_FONT_AXIS_TAG(til::at(axis, 0), til::at(axis, 1), til::at(axis, 2), til::at(axis, 3));
+            _axesVector.emplace_back(DWRITE_FONT_AXIS_VALUE{ dwriteTag, gsl::narrow<float>(value) });

```

Comment:
this `narrow` will throw an exception; are you sure that's what you want?

Reply:
Switched to `narrow_cast`